### PR TITLE
Numbered parser (In Progress)

### PR DESCRIPTION
**Pull request by christinem**
This builds on top of #989.

This PR attempts to create a parser which parses post categories that are listed using a numbered list. Currently it does grab and parse post categories that are listed using a numbered list, but not fully (there are still many bugs), and they are not entered into the database yet.

### CODE REVIEW

**Comment by david-yz-liu** [comment]
@christinem I'm not sure if you're planning on updating this, but it's good to do a merge from master to clean up the diff. (Also there are compiler warnings.)

**Comment by christinem** [comment]
@david-yz-liu I added some more commits to this, and did the merge from master and fixed the warnings. I think this is about it that I'll be adding to this particular PR (unless any fixes are needed of course).

File: app/WebParsing/ParsecCombinators.hs [hunk 1]
```diff
@@ -7,8 +7,9 @@ module WebParsing.ParsecCombinators
      generalCategoryParser,
      parseCategory,
      postInfoParser,
+     parseNumberedLine,
      text, parseAll) where
-
+
```
Comment (david-yz-liu) [inline app/WebParsing/ParsecCombinators.hs]: Delete this blank space.

File: app/WebParsing/PostParser.hs [hunk 2]
```diff
@@ -5,12 +5,13 @@ import qualified Data.Text as T
 import Control.Monad.Trans (liftIO)
 import Text.HTML.TagSoup
 import Text.HTML.TagSoup.Match
+import Data.List.Split
```
Comment (david-yz-liu) [inline app/WebParsing/PostParser.hs]: Don't do unqualified imports (specify which function is being imported).

File: app/WebParsing/PostParser.hs [hunk 3]
```diff
@@ -20,10 +21,12 @@ addPostToDatabase programElements = do
     -- TODO: Remove Focuses from programElements
     let fullPostName = innerText $ take 1 $ filter isTagText programElements
         requirements = last $ sections isRequirementSection programElements
-        liPartitions = partitions isLiTag requirements
+        liPartitions = map parseLi $ partitions isLiTag requirements
+        numberedPartitions = filter (not . T.null) $ map parseNumberedPartition $ getNumberedPartitions requirements
+        nonEmptyPartitions = if (null liPartitions) then numberedPartitions else liPartitions
```
Comment (david-yz-liu) [inline app/WebParsing/PostParser.hs]: Don't use parentheses around the if condition.

File: app/WebParsing/PostParser.hs [hunk 4]
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
...
 
 parseLi :: [Tag T.Text] -> T.Text
 parseLi liPartition = do
     let parsed = P.parse parseCategory failedString (innerText liPartition)
     case parsed of
         Right category -> category
         Left _ -> ""
+
+getNumberedPartitions :: [Tag T.Text] -> [[Tag T.Text]]
+getNumberedPartitions tags = do
```
Comment (david-yz-liu) [inline app/WebParsing/PostParser.hs]: `do` is unnecessary here.

File: app/WebParsing/PostParser.hs [hunk 5]
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
...
+    let pTags = partitions isPTag tags
+        pTagsWithoutBr = map (splitWhen isBRTag) pTags
```
Comment (david-yz-liu) [inline app/WebParsing/PostParser.hs]: You can simplify this by using `concatMap`.

File: app/WebParsing/PostParser.hs [hunk 6]
```diff
@@ -59,16 +62,32 @@ categoryParser tags fullPostName firstCourse liPartitions = do
...
+    concat pTagsWithoutBr
+    where
+        isPTag tag = isTagOpenName "p" tag
```
Comment (david-yz-liu) [inline app/WebParsing/PostParser.hs]: Don't define separate functions; inline them. Note that `isPTag = isTagOpenName "p"`.

**Comment by christinem** [comment]
@david-yz-liu Fixes made.

**Comment by david-yz-liu** [comment]
@christinem sorry this took a while, but thanks!

**Pull request merged**